assets/src/edit-story/app/story/storyProvider.js

```diff
@@ -6,98 +6,111 @@ import PropTypes from 'prop-types';
+	const {
+		selectedElementIds,
+		selectedElements,
+		hasSelection,
+	} = useMemo(
+		() => {
+			if ( ! currentPage ) {
+				return {
+					selectedElements: [],
+					selectedElementIds: [],
+					hasSelection: false,
+				};
+			}
+			const els = currentPage.elements.filter( ( { id } ) => selection.includes( id ) );
+			return {
+				selectedElementIds: selection,
+				selectedElements: els,
+				hasSelection: els.length > 0,
+			};
+		},
+		[ currentPage, selection ],
+	);
 
-	const hasSelection = Boolean( selectedElementIds.length );
+	// This effect loads and initialises the story on first load (when there's no pages).
+	const shouldLoad = pages.length === 0;
+	useLoadStory( { restore, shouldLoad, storyId } );
 
-	const clearSelection = useClearSelection( { selectedElementIds, setSelectedElementIds } );
-	const deleteSelectedElements = useDeleteSelectedElements( { currentPageIndex, pages, selectedElementIds, setPages, setSelectedElementIds } );
-	const setCurrentPageByIndex = useSetCurrentPageByIndex( { clearSelection, setCurrentPageIndex } );
-	const addBlankPage = useAddBlankPage( { pages, setPages, clearSelection } );
-	const deleteCurrentPage = useDeleteCurrentPage( { currentPage, pages, setPages, addBlankPage, setCurrentPageIndex, currentPageIndex } );
-	const selectElementById = useSelectElementById( { setSelectedElementIds } );
-	const toggleElementIdInSelection = useToggleElementIdInSelection( { selectedElementIds, setSelectedElementIds } );
-	const appendElementToCurrentPage = useAppendElementToCurrentPage( { currentPageIndex, pages, setPages, setSelectedElementIds } );
-	const setPropertiesOnSelectedElements = useSetPropertiesOnSelectedElements( { currentPageIndex, pages, selectedElementIds, setPages } );
-	const savePost = useSavePost( { isSaving, storyId, title, postStatus, postAuthor, slug, pages, setLink, setPostStatus, setIsSaving } );
-	const setPropertiesById = useSetPropertiesById( { currentPageIndex, pages, setPages } );
+	// These effects send updates to and restores state from history.
+	useHistoryEntry( { pages, current, selection, story } );
+	useHistoryReplay( { restore } );
 
-	useLoadStory( { storyId, pages, setPages, setTitle, setPostStatus, setPostAuthor, setSlug, setLink, setCurrentPageIndex, clearSelection } );
-	useCurrentPage( { currentPageIndex, pages, setCurrentPage, setCurrentPageNumber } );
-	useHistoryEntry( { currentPageIndex, pages, selectedElementIds } );
-	useHistoryReplay( { setCurrentPageIndex, setPages, setSelectedElementIds } );
-	useSelectedElements( { currentPageIndex, pages, selectedElementIds, setSelectedElements } );
+	// This action allows the user to save the story
+	// (and it will have side-effects, thus the need for `updateStory`)
+	const { updateStory } = api;
+	const saveStory = useSaveStory( { storyId, pages, story, updateStory } );
 
 	const state = {
 		state: {
 			pages,
+			currentPageId,
 			currentPageIndex,
 			currentPageNumber,
 			currentPage,
```

Comment:
^ Any specific reason to not have just one state variable for `currentPage` with params for `id`, `index`, etc.? (Not thinking it would be better otherwise, just curious)

Reply:
I suppose these are needed separately most of the time instead of using/updating all the `currentPage` information?

Reply:
I wanted `currentPage` to be the exact same object also stored in the `pages` array and at the same time I did not want to update index and number on all pages all the time. The reducer could internally manage these, but I didn't find it important. We could however add that quite easily if we wanted to.